### PR TITLE
[flutter_tools] launch devtools from pub

### LINKED ISSUE #65475: [tool_crash] NoSuchMethodError: NoSuchMethodError: The method 'toFilePath' was called on null.Receiver: nullTried calling: toFilePath()
## Command
```
flutter run --flavor dev -t lib/main_dev.dart
```

## Steps to Reproduce
1. ...
2. ...
3. ...

## Logs
NoSuchMethodError: NoSuchMethodError: The method 'toFilePath' was called on null.
Receiver: null
Tried calling: toFilePath()
```
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
#1      defaultHandler (package:devtools_server/src/external_handlers.dart:36:60)
<asynchronous suspension>
#2      serveDevTools (package:devtools_server/src/server.dart:211:21)
#3      ResidentRunner.launchDevTools (package:flutter_tools/src/resident_runner.dart:1195:33)
#4      TerminalHandler._commonTerminalInputHandler (package:flutter_tools/src/resident_runner.dart:1555:32)
#5      TerminalHandler.processTerminalInput (package:flutter_tools/src/resident_runner.dart:1584:13)
#6      _rootRunUnary (dart:async/zone.dart:1198:47)
#7      _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#8      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#9      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#10     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#11     _SyncBroadcastStreamController._sendData (dart:async/broadcast_stream_controller.dart:385:25)
#12     _BroadcastStreamController.add (dart:async/broadcast_stream_controller.dart:250:5)
#13     _AsBroadcastStreamController.add (dart:async/broadcast_stream_controller.dart:484:11)
#14     _rootRunUnary (dart:async/zone.dart:1198:47)
#15     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#16     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#17     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#18     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#19     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:69:11)
#20     _EventSinkWrapper.add (dart:async/stream_transformers.dart:15:11)
#21     _StringAdapterSink.add (dart:convert/string_conversion.dart:238:11)
#22     _StringAdapterSink.addSlice (dart:convert/string_conversion.dart:243:7)
#23     _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:317:20)
```
```
[✓] Flutter (Channel stable, 1.20.2, on Linux, locale en_US.UTF-8)
    • Flutter version 1.20.2 at /home/elementary/.sdk/flutter
    • Framework revision bbfbf1770c (4 weeks ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /home/elementary/.sdk/android
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Android Studio (version 4.0)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.3)
    • VS Code at /usr/share/code
    • Flutter extension version 3.14.1

[✓] Connected device (1 available)
    • Redmi 4A (mobile) • 192.168.43.178:7000 • android-arm64 • Android 7.1.2 (API 25)

• No issues found!

```

## Flutter Application Metadata
**Type**: app
**Version**: 1.1.0+11
**Material**: true
**Android X**: false
**Module**: false
**Plugin**: false
**Android package**: null
**iOS bundle identifier**: null
**Creation channel**: stable
**Creation framework version**: 68587a0916366e9512a78df22c44163d041dd5f3
### Plugins
background_fetch-0.5.6
camera-0.5.8+5
device_info-0.4.2+7
firebase_auth-0.16.1
firebase_auth_web-0.1.3+1
firebase_core-0.4.5
firebase_core_web-0.1.1+2
firebase_crashlytics-0.1.4+1
firebase_messaging-6.0.16
flutter_image_compress-0.7.0
flutter_local_notifications-0.8.4+3
flutter_plugin_android_lifecycle-1.0.8
fluttertoast-4.0.1
google_sign_in-4.5.3
google_sign_in_web-0.9.1+1
image_picker-0.6.7+7
location-3.0.2
location_web-1.0.0
package_info-0.4.3
path_provider-1.6.14
path_provider_linux-0.0.1+2
path_provider_macos-0.0.4+3
permission_handler-5.0.1+1
searchable_dropdown-1.1.3
share-0.6.5
shared_preferences-0.5.10
shared_preferences_linux-0.0.2+2
shared_preferences_macos-0.0.1+10
shared_preferences_web-0.1.2+7
sqflite-1.3.1+1
url_launcher-4.2.0+3

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

https://github.com/flutter/flutter/issues/65128

Fixes https://github.com/flutter/flutter/issues/65475

### CODE REVIEW

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -1704,43 +1705,76 @@ String nextPlatform(String currentPlatform, FeatureFlags featureFlags) {
...
+    if (_devtoolsUri == null) {
       return null;
     }
+    return DevToolsServerAddress(_devtoolsUri.host, _devtoolsUri.port);
   }
 
   Future<void> close() async {
-    await _devtoolsServer?.close();
-    _devtoolsServer = null;
+    _devtoolsProcess.kill();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/resident_runner.dart:1773]: This does not seem to kill the devtools server, i would be worried about leaking it. FYI @jacob314

Comment (jacob314) [reply packages/flutter_tools/lib/src/resident_runner.dart:1773]: we can send a command to the devtools server on stdout asking it to exit. That would be a couple line change to devtools.

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 2]
```diff
@@ -1704,43 +1705,76 @@ String nextPlatform(String currentPlatform, FeatureFlags featureFlags) {
...
-      await devtools_server.launchDevTools(
-        <String, dynamic>{
-          'reuseWindows': true,
-        },
-        observatoryAddress,
-        'http://${_devtoolsServer.address.host}:${_devtoolsServer.port}',
-        false,  // headless mode,
-        false,  // machine mode
-      );
+      _devtoolsProcess = await _processManager.start(<String>[
```
Comment (jacob314) [inline packages/flutter_tools/lib/src/resident_runner.dart:1733]: there is a daemon mode for devtools inspired by the one in flutter tools. Pass it in and you won't need regex match to detect what port is used.
https://github.com/flutter/devtools/blob/master/packages/devtools_server/lib/src/server.dart#L85

**Comment by jonahwilliams** [comment]
I don't have more time to look at this

**Pull request closed without merging**